### PR TITLE
Raise FlowProjectDefinitionError for FlowProject definition related errors

### PR DESCRIPTION
**Pull request by kidrahahjo**
## Description

Errors raised in signac-flow when there is a flaw in FlowProject definition were not consistent. 
This pull request introduces a custom error, `FlowProjectDefinitionError` to handle flaws in FlowProject definition.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

File: flow/directives.py [hunk 1]
```diff
@@ -144,7 +144,9 @@ def _add_directive(self, directive):
                 f"Expected a _Directive object. Received {type(directive)}."
             )
         if directive._name in self._directive_definitions:
-            raise ValueError(f"Cannot redefine directive name {directive._name}.")
+            raise FlowProjectDefinitionError(
+                f"Cannot redefine directive name {directive._name}."
+            )
```
Comment (b-butler) [inline flow/directives.py:149]: This method is private (underscored) and not meant to be called by the user. This is more to ensure that we don't violate this when we define an environment's default directives. We can revert this change.

File: flow/project.py [hunk 2]
```diff
@@ -499,7 +500,7 @@ def _eligible(self, aggregate, ignore_conditions=IgnoreConditions.NONE):
 
         """
         if not isinstance(ignore_conditions, IgnoreConditions):
-            raise ValueError(
+            raise FlowProjectDefinitionError(
```
Comment (b-butler) [inline flow/project.py:503]: This is not a definition error.

File: flow/project.py [hunk 3]
```diff
@@ -3236,7 +3239,7 @@ def run(
             num = None
 
         if not isinstance(ignore_conditions, IgnoreConditions):
-            raise ValueError(
+            raise FlowProjectDefinitionError(
```
Comment (b-butler) [inline flow/project.py:3242]: Not a definition error.

**Review by b-butler — CHANGES_REQUESTED** [review]
There do not seem to be places where the new error needs to be added, but there are some changes that need to be reverted.

**Review by b-butler — APPROVED** [review]
LGTM, I had checked the repo for places where we raise exceptions and I think this covers all the changes to the new error type. I think the code coverage can be ignored here, and potentially addressed in a future PR, since we already doe not test these errors.

**Comment by kidrahahjo** [comment]
@b-butler I updated the changelog, please review the changelog so that we can merge this PR.

**Review by b-butler — CHANGES_REQUESTED** [review]
@kidrahahjo add a line to `Added` as well, saying something like `Add ``FlowProjectDefinitionError`` for exceptions involving ``FlowProject`` definitions.

File: flow/errors.py [hunk 4]
```diff
@@ -61,3 +61,9 @@ def parse(self, parser):
     def err(self, msg, caller):
         """Raise a template error."""
         raise jinja2.TemplateError(msg)
+
+
+class FlowProjectDefinitionError(ValueError):
+    """Indicates that the project was incorrectly defined."""
```
Comment (bdice) [inline flow/errors.py:67]:
```suggestion
    """Indicates an invalid FlowProject definition."""
```

File: flow/project.py [hunk 5]
```diff
@@ -676,8 +677,8 @@ def __call__(self, func):
         """
         if hasattr(func, "_flow_groups"):
             if self.name in func._flow_groups:
-                raise ValueError(
-                    f"Cannot register existing name {func} with group {self.name}"
+                raise FlowProjectDefinitionError(
+                    f"Cannot reregister operation {func} with the group, {self.name}"
```
Comment (bdice) [inline flow/project.py:679]:
```suggestion
                    f"Cannot reregister operation '{func}' with the group '{self.name}'."
```

**Comment by bdice** [comment]
> 
> 
> @kidrahahjo add a line to `Added` as well, saying something like `Add `FlowProjectDefinitionError` for exceptions involving ``FlowProject`` definitions.

This suggestion seems very similar to the existing entry in the "Changed" section. @b-butler Are you saying that we need to duplicate the message here?

Also I'm going to push a couple changes to this PR that add test coverage for some errors that aren't being reached.

File: flow/project.py [hunk 6]
```diff
@@ -4257,11 +4262,13 @@ def foo(job):
...
-            raise ValueError(f"Repeat definition of group with name '{name}'.")
+            raise FlowProjectDefinitionError(
+                f"Repeat definition of group with name '{name}'."
+            )
         if any(
             name == operation_name for operation_name, _ in cls._OPERATION_FUNCTIONS
         ):
-            raise ValueError(
+            raise FlowProjectDefinitionError(
                 f"Cannot create a group with the same name as the existing operation {name}"
```
Comment (bdice) [inline flow/project.py:4256]:
```suggestion
                f"Cannot create a group with the same name as the existing operation '{name}'."
```

File: flow/project.py [hunk 7]
```diff
@@ -687,7 +688,7 @@ def __call__(self, func):
     def _set_directives(self, func, directives):
         if hasattr(func, "_flow_group_operation_directives"):
             if self.name in func._flow_group_operation_directives:
-                raise ValueError(
+                raise FlowProjectDefinitionError(
                     f"Cannot set directives because directives already exist "
                     f"for {func} in group {self.name}"
```
Comment (bdice) [inline flow/project.py:691]:
```suggestion
                    "Cannot set directives because directives already exist "
                    f"for operation '{func}' in group '{self.name}'."
```

File: flow/operations.py [hunk 8]
```diff
@@ -42,7 +43,7 @@ def hello(job):
         prefixes to the shell command provided here.
     """
     if getattr(func, "_flow_with_job", False):
-        raise RuntimeError(
+        raise FlowProjectDefinitionError(
             "@cmd should appear below the @with_job decorator in your script"
```
Comment (bdice) [inline flow/operations.py:47]:
```suggestion
            "The @flow.cmd decorator must appear below the @flow.with_job decorator."
```

File: flow/aggregates.py [hunk 9]
```diff
@@ -347,12 +349,12 @@ def __call__(self, func=None):
...
         if not callable(func):
-            raise TypeError(
+            raise FlowProjectDefinitionError(
                 "Invalid argument passed while calling the aggregate "
                 f"instance. Expected a callable, got {type(func)}."
             )
         if getattr(func, "_flow_with_job", False):
-            raise RuntimeError(
+            raise FlowProjectDefinitionError(
                 "The @with_job decorator cannot be used with aggregation."
```
Comment (bdice) [inline flow/aggregates.py:358]:
```suggestion
                "The @flow.with_job decorator cannot be used with aggregation."
```

**Review by bdice — COMMENTED** [review]
I rewrote some of the error messages for consistency. I'm going to merge these changes.

**Review by bdice — APPROVED** [review]
This looks good to me. @b-butler please merge once you are satisfied with your requests about the changelog. I think it is fine as-is.

**Review by b-butler — APPROVED** [review]
:+1:
